pkg/ring/kvstore.go

```diff
@@ -18,35 +19,55 @@ var inmemoryStore KVClient
 type KVClient interface {
 	CAS(ctx context.Context, key string, f CASCallback) error
 	WatchKey(ctx context.Context, key string, f func(interface{}) bool)
+	WatchPrefix(ctx context.Context, prefix string, f func(interface{}) bool)
 	Get(ctx context.Context, key string) (interface{}, error)
 	PutBytes(ctx context.Context, key string, buf []byte) error
 }
 
+// KVConfig is config for a KVStore currently used by ring and HA tracker,
+// where store can be consul or inmemory.
+type KVConfig struct {
+	Store  string       `yaml:"store,omitempty"`
+	Consul ConsulConfig `yaml:"consul,omitempty"`
+}
+
+// RegisterFlags adds the flags required to config this to the given FlagSet
+func (cfg *KVConfig) RegisterFlags(f *flag.FlagSet, prefix string) {
+	cfg.Consul.RegisterFlags(f, prefix)
+	if prefix == "" {
+		prefix = "ring"
+	}
+	prefix = prefix + "."
+	f.StringVar(&cfg.Store, prefix+"store", "consul", "Backend storage to use for the ring (consul, inmemory).")
+}
+
 // CASCallback is the type of the callback to CAS.  If err is nil, out must be non-nil.
 type CASCallback func(in interface{}) (out interface{}, retry bool, err error)
 
-func newKVStore(cfg Config) (KVClient, error) {
+// NewKVStore creates a new KVstore client (inmemory or consul) based on the config,
+// encodes and decodes data for storage using the codec.
+func NewKVStore(cfg Config, codec Codec) (KVClient, error) {
```

Comment:
This should take a `KVConfig`, no?

Reply:
I think so, but the ring config is what contains the `Mock` value. Where should we move that?

Reply:
I'd move the mock in to KVConfig too.